ee/tabby-webserver/src/service/github_repository_provider.rs

```diff
@@ -113,6 +117,83 @@ impl GithubRepositoryProviderService for GithubRepositoryProviderServiceImpl {
 
         Ok(urls)
     }
+
+    async fn refresh_repositories(&self) -> Result<()> {
+        let mut cached_repositories: HashSet<_> = self
+            .list_github_provided_repositories_by_provider(vec![], None, None, None, None)
+            .await?
+            .into_iter()
+            .map(|repo| repo.vendor_id)
+            .collect();
+
+        for provider in self
+            .list_github_repository_providers(None, None, None, None)
+            .await?
+        {
+            let repos = fetch_all_repos(&provider).await?;
+            for repo in repos {
+                let id = repo.id.to_string();
+                let Some(url) = repo.git_url.map(|url| url.to_string()) else {
+                    continue;
+                };
+                // Remove IDs as we process them so the remaining IDs are ones that were not found in the listing
+                if cached_repositories.remove(&id) {
+                    self.db
+                        .update_github_provided_repository(id, repo.name, url)
+                        .await?;
+                } else {
+                    self.db
+                        .create_github_provided_repository(
+                            provider.id.clone().as_rowid()?,
+                            id,
+                            repo.name,
+                            url,
+                        )
+                        .await?;
+                }
+            }
+        }
+
+        // Clean up repositories which were not returned by any listing
+        for removed_repository in cached_repositories {
```

Comment:
consider implement the functionality with sql + updated_at.

`DELETED rows WHERE updated_at < X`

You might also added `created_at` and `updated_at` column to the table

Reply:
I believe this is better for performance, more readable, and less complex. If adding `created_at` and `updated_at` we would then also want an index on `(vendor_id, created_at, updated_at)` and I believe that's more error-prone than simply removing any IDs which were not returned

Reply:
there's no need to index created at / updated at, scan performance is sufficient. This mechanism is also more robust (`DELETE ... WHERE` is transactional)